src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletActor.scala

```diff
@@ -0,0 +1,202 @@
+
+  private val registry = new Registry
+
+  //todo: pass as parameter, add to config
+  val boxSelector: BoxSelector = DefaultBoxSelector
+
+  private val prover = new ErgoProvingInterpreter(seed)
+
+  private var height = 0
+  private var lastBlockUtxoRootHash = ADDigest @@ Array.fill(32)(0: Byte)
+
+  private val trackedAddresses: mutable.Buffer[ErgoAddress] =
+    mutable.Buffer(prover.dlogPubkeys: _ *).map(P2PKAddress.apply)
+
+  private val trackedBytes: mutable.Buffer[Array[Byte]] = trackedAddresses.map(_.contentBytes)
+
+  //todo: make resolveUncertainty(boxId, witness)
+  private def resolveUncertainty(): Unit = {
+    registry.nextUncertain().foreach { uncertainBox =>
+      val box = uncertainBox.box
+
+      val lastUtxoDigest = AvlTreeData(lastBlockUtxoRootHash, 32)
+
+      val testingTx = UnsignedErgoLikeTransaction(
+        IndexedSeq(new UnsignedInput(box.id)),
+        IndexedSeq(new ErgoBoxCandidate(1L, Values.TrueLeaf))
+      )
+
+      val context =
+        ErgoLikeContext(height + 1, lastUtxoDigest, IndexedSeq(box), testingTx, box, ContextExtension.empty)
+
+      prover.prove(box.proposition, context, testingTx.messageToSign) match {
+        case Success(_) =>
+          log.info(s"Uncertain box is mine! $uncertainBox")
+          registry.makeTransition(uncertainBox, uncertainBox.makeCertain())
+        case Failure(_) =>
+        //todo: remove after some time? remove spent after some time?
+      }
+    }
+  }
+
+  def scan(tx: ErgoTransaction, heightOpt: Option[Height]): Boolean = {
+    tx.inputs.foreach { inp =>
+      val boxId = ByteArrayWrapper(inp.boxId)
+      if (registry.registryContains(boxId)) {
+        registry.makeTransition(boxId, ProcessSpending(tx, heightOpt))
+      }
+    }
+
+    tx.outputCandidates.zipWithIndex.exists { case (outCandidate, outIndex) =>
+      trackedBytes.find(t => outCandidate.propositionBytes.containsSlice(t)) match {
+        case Some(_) =>
+          val idxShort = outIndex.toShort
+          val box = outCandidate.toBox(tx.serializedId, idxShort)
+          val bu = heightOpt match {
+            case Some(h) => UnspentOnchainBox(tx, idxShort, h, box, Uncertain)
+            case None => UnspentOffchainBox(tx, idxShort, box, Uncertain)
+          }
+          bu.register(registry)
+          true
+        case None =>
+          false
+      }
+    }
+  }
+
+  private def extractFromBlock(fb: ErgoFullBlock): Int = {
+    height = fb.header.height
+    lastBlockUtxoRootHash = fb.header.stateRoot
+    fb.transactions.count(tx => scan(tx, Some(height)))
+  }
+
+  def scanLogic: Receive = {
+    case ScanOffchain(tx) =>
+      if (scan(tx, None)) {
+        self ! Resolve
+      }
+
+    case Resolve =>
+      resolveUncertainty()
+      //todo: avoid magic number, use non-default executor? check that resolve is not scheduled already
+      if (registry.uncertainBoxes.nonEmpty) {
+        context.system.scheduler.scheduleOnce(10.seconds)(self ! Resolve)
```

Comment:
Here we send twice more `Resolve` messages than was sent after scanning transactions